### ISSUE TITLE
[Idea/Enhancement] Use Docker secrets for certificates

### ISSUE THREAD

**Issue by morph027** [containous/traefik#1619]
### Do you want to request a *feature* or report a *bug*?

feature / idea / enhancement

### What did you do?

I've started using [Docker Flow Proxy](http://proxy.dockerflow.com/) before i stumbled upon your great piece of cheese ;) I'd like the simplicity of traefik for webapps (most of my/our apps). For our production swarm, we wanted to enroll SSL, which is a bit more fun with [Docker Flow Proxy: Adding Certificates As Docker Secrets](http://proxy.dockerflow.com/certs/#adding-certificates-as-docker-secrets) as i do not need to maintain any configuration file or backend storage for certs.

### What did you expect to see?

Be able to use certs from Docker secrets. Bonus would be frontend configuration via commandline/labels instead of configuration file (All knowledge to swarm ;) )

**Comment by denzuko** [comment]
👍 on this.

Or at the least update the documentation for real world certificate deployments. Internal tools on Class A/B/C networks which that cannot reach an ASN routed broadcast more than obviously *cannot* reach back to `acme-v01.api.letsencrypt.org`. Thus require the use of internal PKI and administratively signed certs.

Documentation could include directives for how to apply docker labels to above use case.

**Comment by bgerstle-lookback** [comment]
Seems like this would require (or be related to) using docker secrets as the [key value store](https://docs.traefik.io/user-guide/kv-config/#store-configuration-in-key-value-store). We're trying to learn more about the [clustering/HA feature](https://docs.traefik.io/user-guide/cluster/) and how it relates to the KV store, but documentation is a bit sparse. The main problems we're trying to solve are: rolling out new Traefik machines/instances (i.e. blue/green), disaster recovery.

**Comment by bgerstle** [comment]
Seems like this would require (or be related to) using docker secrets as the [key value store](https://docs.traefik.io/user-guide/kv-config/#store-configuration-in-key-value-store). We're trying to learn more about the [clustering/HA feature](https://docs.traefik.io/user-guide/cluster/) and how it relates to the KV store, but documentation is a bit sparse. The main problems we're trying to solve are: rolling out new Traefik machines/instances (i.e. blue/green), disaster recovery.

**Comment by denzuko** [comment]
@bgerstle Yes and No, yes in that its closely working with the same concepts. No in that I'm trying to use pre generated certificates stored in docker secrets (ie encrypted in ram version of bind mount) which overlay a file into the container's filesystem.

This is because our security policies do not allow for internal applications being on the public internet, a requirement for acme/letsencrypt. Nor are we looking to introduce additional complexity with consul/etcd when docker themselves have implemented a RAFT K/V store internally to Moby.

There's some discussion on slack - https://traefik.slack.com/messages/C0CDT22PJ/search/%3C@U7BJ324UF%7C@Dwight%20S.%3E/

**Comment by lucj** [comment]
Any update on this ?

**Comment by nathanaherne** [comment]
I would like to know more about this feature request and whether it is on the development roadmap? 

It would be great to be able to use Docker secrets to achieve HA. It seems silly to use an external piece of software, when the tools are there natively..... Unless I am missing something.

**Comment by Mika56** [comment]
Just so you know, I've tried to implement this and as far as I can tell, updating a secret is not possible. You can only create and delete secrets.
Adding or removing a secret to a running container/service will restart it.

In order to use secrets for certificates, you would need to use them in a way they were not meant to be

**Comment by morph027** [comment]
Ah, that's right ... you can use rolling secrets with mappings, but this needs a service restart.

In this case, this would cause a traefik restart, which should be okay if you scale them accross the cluster and restart all of them in proper order so that at least one is running ;)

However, i thought of this as a nice-to-have. Spinning up an etcd or whatever cluster inside swarm is no magic, but sometimes i like to keep the amount of things which can break as minimal as possible ;)

**Comment by Mika56** [comment]
I couldn't agree more, and in my experience it's always the KV store that fails and make Traefik stop working.

You could always use a second service dedicated to managing certificates, but as I've stated, that's using secrets in a way they were not meant to be, plus you would probably need all Traefik instances to be running on manager nodes to be able to access the secret (and managing a second service doesn't seem to simplify things)

**Comment by morph027** [comment]
Jip. Looks like your #2855 is the better way to go :+1: (I've already played with this a while ago using a custom entrypoint script which retrieves all the secrets ans then starts traefik. but any change there requires a restart too)

**Comment by Mika56** [comment]
For my use case, Vault is for an other usage.

Spinning up a Vault instance that stores to Consul and provides certificates to Traefik isn't any simpler than spinning up a Consul instance that directly stores certificates for Traefik.\
It would be nice if we could find a way for Traefik to manage the certificates on itself, without any other software

**Comment by nathanaherne** [comment]
It seems there is quite a lively discussion about being able to update configs and secret.

https://github.com/moby/moby/issues/29882

**Comment by denzuko** [comment]
Seems to be we're all agreeing to here; the lease amount of moving parts and not depending on external tools for key stores.

Now after reading over the thread on moby and how they're approaching [secrets/configs](https://github.com/moby/moby/issues/35048#issuecomment-333778750). It would appear they're trying to go the route of a merkle tree style immutable objects with rolling over to new container iterations.

While the moby team works out how containers should handle the behavior of updated secrets. My question here becomes less on how to manage traefik with updating certificates and more towards how could one approach using docker secrets for deploying an existing certificate in relation to say treating it as a flat file inside the traefik container?

**Comment by morph027** [comment]
Not sure if i got you right, but secrets are present as flat files in `/run/secrets/...`

**Comment by Mika56** [comment]
They are read-only in /run/secrets/

**Comment by kevb** [comment]
I'm in the same boat as others - using a KV store introduces too much complexity, storing certificates in a volume ties me to undesirable patterns (single instance of Traefik or tied to a specific node or both). Unfortunately I don't see Secrets being a solution any time soon.

In the meantime I'm just terminating SSL at an external Nginx load balancer and managing certificates there...

Are there any other possible solutions? I noticed #1620 suggestion to support s3, and while I agree this probably shouldn't be part of Traefik, perhaps a plugin system?

**Comment by denzuko** [comment]
@kevb wasn't that what docker itself did in the [s3 storage driver](https://docs.docker.com/registry/storage-drivers/s3/)?

**Comment by kevb** [comment]
@denzuko I believe this is a storage driver for docker registry images rather than for arbitrary storage.

**Comment by SMARoovers** [comment]
I would suggest to change this request or create a new one to focus on the bonus part:

`Bonus would be frontend configuration via commandline/labels instead of configuration file`

Because as stated by morph027 secrets are available in the container under /run/secrets and could therefore be used as a path to the certfile.

E.g.
`      labels:
        - "traefik.port=5000"
        - "traefik.frontend.rule=Host:domain.local"
        - "traefik.docker.network=traefik-network"
        - "traefik.enable=true"
        - "traefik.frontend.entrypoints.https.tls.certificates.certfile=/var/run/secrets/domain.local.cert"
        - "traefik.frontend.entrypoints.https.tls.certificates.keyfile=/var/run/secrets/domain.local.key"
        - "traefik.frontend.entryPoints=http,https"`

**Comment by shawngmc** [comment]
There are also cases where you need other values that would be best kept as secrets. For example, I use Traefik's lets' encrypt/route-53 support - so while the certs are dynamic and I really don't care, the AWS auth info should be a secret.

**Comment by lwl12** [comment]
Key rotate should exactly useful, and may especially important for let's encrypt users...

I'll keep following this.

**Comment by RamjiVE** [comment]
Any update on this??

**Comment by EthraZa** [comment]
Searching the possiblity of use secrets to store certs I came here. Any news?

**Comment by Aster-the-Med-Stu** [comment]
Just a side note here:

Using docker secret would be a headache when it comes to updating certificate. In https://github.com/moby/moby/issues/29882, there've already been many people complaining about this.
